connectors/sources/onedrive.py

```diff
@@ -637,24 +759,35 @@ async def get_docs(self, filtering=None):
                     async for entity, download_url in self.client.get_owned_files(
                         user_id, skipped_extensions, pattern
                     ):
-                        entity = self.prepare_doc(entity)
-                        if entity["type"] == FILE and download_url:
-                            yield entity, partial(
-                                self.get_content, entity.copy(), download_url
-                            )
-                        else:
-                            yield entity, None
+                        yield self.send_document_to_es(entity, download_url)
         else:
-            async for user in self.client.list_users():
-                user_id = user.get("id")
-
-                async for entity, download_url in self.client.get_owned_files(user_id):
-                    entity = self.prepare_doc(entity)
-                    if entity["type"] == FILE and download_url:
-                        yield await self._decorate_with_access_control(
-                            entity, user_id, entity.get("_id")
-                        ), partial(self.get_content, entity.copy(), download_url)
-                    else:
-                        yield await self._decorate_with_access_control(
-                            entity, user_id, entity.get("_id")
-                        ), None
+            async for users in self.client.list_users():
```

Comment:
Can you do something like

```python
requests = [build_owned_files_url(user.id) async for user in self.client.list_users()]

# json_batching should retry for failed requests
async for response in self.json_batching(requests):
	# process response 
```

This can also be applied to the `if` branch (when advanced rules are configured) with some tweaking.

Reply:
We need to pass batches of users for building requests so we need to use `iterable_batches_generator`.  

Also, i believe we should not use batching in advanced rules since in a given rule, the length of `owners` list can be in single digit, so using batching might be inefficient in such cases.

Reply:
>We need to pass batches of users for building requests

I don't quite understand this.

I imagine the `json_batching` is agnostic of what kind of requests it's going to process. It will simply accept a list of requests, and try to batch them as much as it can. Pseudocode of  `json_batching`:
```python
async def json_batching(self, requests):
	if not requests:
		return
	for request in requests:
		request["retry_count"] = 0

	batch_url = parse.urljoin(BASE_URL, ENDPOINTS[BATCH])

    while requests:
        requests_to_process = pop_first_20_requests(requests)
        
        async with self.client.post(batch_url, requests_to_process) as batch_response:
            batch_response = await batch_response.json()

        	for response in batch_response.get("responses", []):
        		if response.get("status", 200) == 200:
        			if has_next_page(response):
        				requests.append(next_page_request(response))
        			yield response
        	else:
        		request = look_up_request_by_id(requests_to_process, response.get("id"))
        		request["retry_count"] += 1
        		if request["retry_count"] > 3:
        			logger.error("Request {request} failed after 3 retries, giving up...")
        		else:
        			requests.append(request)
```

Reply:
WDYT if we revert the changes to `list_users` here:

https://github.com/elastic/connectors/blob/0fb6efde111e3543f125fef6b64d42b9a9dbcd48/connectors/sources/onedrive.py#L388-L389

I'd expect a method `list_users` will yield a user object.

And we could build `requests` for all users, instead of only a batch of users, and pass `requests` to `json_batching`, which is more efficient.

Reply:
This can be done but I think, if we build all the `requests` and then move towards `json_batching` what happens is that the execution would wait till all the requests are added, effectively the second `async for loop` won't execute until the first one is completed. 
```python
requests = []
async for user in self.client.list_users():
    requests.append(self.build_owned_files_url(users))

async for response in self.json_batching(batched_apis=requests):
    files = response.get("body", {}).get("value", [])
```